presto-main/src/main/java/com/facebook/presto/execution/buffer/ArbitraryOutputBuffer.java

```diff
@@ -219,10 +226,23 @@ public void enqueue(List<SerializedPage> pages)
         long rowCount = pages.stream().mapToLong(SerializedPage::getPositionCount).sum();
         totalRowsAdded.addAndGet(rowCount);
         totalPagesAdded.addAndGet(pages.size());
+        outstandingPagePerLifespan.computeIfAbsent(lifespan, ignore -> new AtomicLong()).addAndGet(pages.size());
 
         // create page reference counts with an initial single reference
         List<SerializedPageReference> serializedPageReferences = pages.stream()
-                .map(pageSplit -> new SerializedPageReference(pageSplit, 1, () -> memoryManager.updateMemoryUsage(-pageSplit.getRetainedSizeInBytes())))
+                .map(pageSplit -> new SerializedPageReference(
+                        pageSplit,
+                        1,
+                        () -> {
+                            long oustandingPage = outstandingPagePerLifespan.get(lifespan).decrementAndGet();
+                            Consumer<Lifespan> callback = lifespanFinishedCallback;
+                            if (oustandingPage == 0 && callback != null
```

Comment:
Can `callback` be null here? If yes, this means it's optional and we can use `Optional` for that. Otherwise, we can just assert that it has to be non-null here and call it.